### ISSUE TITLE
RAG: Research M2 reranker model for query-result relevance (Phase 3)

### ISSUE THREAD

**Issue by marctjones** [marctjones/klareco#534]
## Problem

After implementing deterministic solutions (#532, #533), we may still need learned reranking for complex semantic matching.

**Example:** "Kiu fondis Esperanton?"
- Both "Zamenhof founded Esperanto" and "Schmidt co-founded Esperanto movement" are:
  - ✓ Grammatically plausible (M1 scores both high)
  - ✓ Same entity types (PERSON founded LANGUAGE)
  - ❓ But only one directly answers the query

Need to score **query-result relevance**, not just plausibility.

## Proposed Solution

Build M2 Reranker - a learned model that scores how well a result answers a query.

### Architecture

```python
class M2Reranker(nn.Module):
    """Scores query-result relevance."""
    
    def forward(self, query_embs, result_embs):
        # Input: query AST embeddings (S-V-O)
        #        result AST embeddings (S-V-O)
        # Output: relevance_score (0-1)
```

**Parameters:** 20-100M (trainable on laptop, fits vision)

### Training Data

Generate Q&A pairs from corpus:

```python
# Positive: (question, correct_answer)
query = {verb: 'fond', object: 'esperant', subject: None}
answer = {verb: 'fond', object: 'esperant', subject: 'zamenhof'}
label = 1

# Negative: (question, wrong_answer)
wrong = {verb: 'fond', object: 'esperant', subject: 'schmidt'}
label = 0
```

### Integration

Update RAG pipeline:
1. Retrieve candidates (Kuzu)
2. M1 filter (grammatical plausibility)
3. **M2 rerank** (query-result relevance) ← NEW
4. Return top-k

## Design Alignment

✅ **Explicitly allowed** - CLAUDE.md lists "retrieval reranking" as minimal learned component  
✅ **Decomposable** - Can measure: "Retrieval: 60%, M1: 20%, M2: 20%"  
✅ **Small model** - 20-100M params fits "trainable on laptop" goal  
⚠️ **Phase 3 only** - Implement AFTER proving deterministic solutions (#532, #533)  

## Empirical Validation

This is a **thesis test**:

> Can we solve query disambiguation with deterministic rules, or do we need learned reranking?

**The Experiment:**
1. ✅ Phase 1: Compound filtering (#532)
2. ✅ Phase 2: Entity classification (#533)
3. 📊 **Measure**: Do Solutions 1+2 solve 90%+ of cases?
4. ❓ Phase 3: Only add M2 if <90% solved

From VISION.md:
> "If learned layers contribute <5%: deterministic is sufficient"
> "If learned layers contribute 10-15%: hybrid approach optimal"

## Expected Impact

**Effort:** 2-3 months  
**Impact:** Solves 95%+ of cases if deterministic solutions insufficient  
**Requires:**  
- Q&A training dataset generation
- Model training pipeline
- Integration into RAG

## Priority

**Low - Research only**

⚠️ **DO NOT implement unless:**
1. Solutions #532 and #533 are complete
2. Measurement shows <90% disambiguation accuracy
3. Evidence proves learned component is needed

## Files to Create

- `klareco/models/m2_reranker.py` - Model architecture
- `scripts/generate_m2_training_data.py` - Q&A pair generation
- `scripts/train_m2_reranker.py` - Training script
- `tests/test_m2_reranker.py` - Tests

## Related

- Issue #532 - Compound filtering (prerequisite)
- Issue #533 - Entity classification (prerequisite)
- VISION.md - Validates "decomposable contributions" thesis

**Comment by marctjones** [comment]
Closing as part of a 2026-09-08 backlog cleanup (see PLAN.md, DESIGN.md, CLAUDE.md's merge gate).

This issue predates the merge gate (adopted 2026-07-13) and the AST-native orchestrator rewrite. It describes work against the pre-orchestrator RAG pipeline or the learned model stack (M1/M2 embeddings, compositional embeddings, entity classifiers, neural rerankers) that was pruned from the repo -- klareco/embeddings/, klareco/models/, and klareco/summarization/ no longer exist. It has no milestone and was never benchmarked, matching exactly the failure mode CLAUDE.md documents (25 capability issues filed 2026-05-26, 8 implemented, none benchmarked).

Labeled archived-2026-09 so this batch is recoverable: gh issue list --state closed --label archived-2026-09. If this capability is still wanted, re-file it against the current architecture naming the benchmark metric and test the merge gate requires.
